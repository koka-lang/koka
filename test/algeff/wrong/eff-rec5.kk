
effect state {
  fun get() : hide
  fun set(x : hide) : ()
}
type hide {
  Hide(unhide : int -> state int)
}

val new-ref = handler(s) {
  get() -> resume(s,s)
  set(x) -> resume((),x)
}

fun recurse() {
  using new-ref(Hide(id))
  fun foo(i : int) {
    (get().unhide)(i)
  }
  set(Hide(foo))
  foo(42)
}

type xstate :: HX {
  con OpsSet( xset : xset )
}

type xset {
  con OpSet( x : xhide )
}

type xhide {
  con XHide( unhide : int -> xstate int )
}
