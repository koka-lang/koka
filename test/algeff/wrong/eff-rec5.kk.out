test/algeff/wrong/eff-rec5.kk(32, 1): error: Type xhide is declared as being (co)inductive but it occurs
 recursively in a negative position.
 hint: declare it as a 'type rec' (or 'effect rec)' to allow negative occurrences